payments/models.py

```diff
@@ -154,3 +169,345 @@ def __str__(self):
+                )
+
+        # Check that it has either product or price
+        if all([not self.product, not self._product_object_id, not self.price]):
+            raise ValidationError(
+                _("Order must have either product object or price value")
+            )
+
+        if self._state.adding:
+            # Don't allow creating with product and price
+            if (
+                self.price
+                and self.tax_percentage
+                and (self.product or self._product_object_id)
+            ):
+                raise ValidationError(
+                    _("Cannot receive product and plain price/tax_percentage")
+                )
+        else:
+            old_instance = Order.objects.get(id=self.id)
+            # If the product is being changed
+            if old_instance.product != self.product:
+                raise ValidationError(
+                    _("Cannot change the product assigned to this order")
+                )
+
+            # If the lease is being changed
+            if old_instance.lease and self.lease != old_instance.lease:
+                raise ValidationError(
+                    _("Cannot change the lease associated to this order")
+                )
+
+    def is_full_season(self):
+        return (
+            self.lease.start_date == calculate_berth_lease_start_date()
+            and self.lease.end_date == calculate_berth_lease_end_date()
+        )
+
+    def _create_order_lines(self, section):
+        for service in ProductServiceType.FIXED_SERVICES():
+            # If the section has the service (prop) and it's valid (True)
+            if hasattr(section, service.name.lower()) and getattr(
+                section, service.name.lower()
+            ):
+                # Retrieve the product for that service
+                product = AdditionalProduct.objects.get(
+                    service=service, period=PeriodType.SEASON
+                )
+
+                OrderLine.objects.create(
+                    order=self, product=product,
+                )
+
+    def save(self, *args, **kwargs):  # noqa: C901  is too complex
+        self.full_clean()
+
+        # If the product is being added from the admin (only the ID is passed)
+        if self._product_object_id and not self.product:
+            # Try to get a BerthProduct (BP)
+            product = BerthProduct.objects.filter(id=self._product_object_id).first()
+            # If the BP was not found, try getting a WinterStorageProduct
+            product = (
+                product
+                if product
+                else WinterStorageProduct.objects.filter(
+                    id=self._product_object_id
+                ).first()
+            )
+            # If for some reason neither was found (shouldn't be the case), raise an error
+            if not product:
+                raise ValidationError(_("The product passed is not valid"))
+
+            self.product = product
```

Comment:
maybe this could be encapsulated in a helper function `self._assign_product_from_object_id()`? 🤔  would clean up `save` a bit